### ISSUE TITLE
Parsing: sqlfluff parser does not detect extra comma as incorrect syntax with multiple cte's

### ISSUE THREAD

**Issue by jcocozza** [sqlfluff/sqlfluff#6426]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

I was writing a query of the following variety for hive sql.

```sql
WITH cte AS (
    SELECT
        a,
        b
    FROM foo
),

cte2 AS (
    SELECT
        a,
        c
    FROM bar
),

SELECT
    cte.a,
    cte.b,
    cte2.c
FROM cte
LEFT JOIN cte2
    ON cte.a = cte2.a;
```

Those with a keen eye will notice a small error in the second CTE. 
There should not be a final comma:
```sql
cte2 AS (
    SELECT
        a,
        c
    FROM bar
)
```

I know that hive will be unable to run a query with this kind of syntax error. The ansi parser also does not get upset, but I don't know if that should be the case or not.

### Expected Behaviour

The parser should get upset.

### Observed Behaviour

No behavior to observe, which is the problem. The parser should detect the incorrect sql.

### How to reproduce

```sql
-- test.sql
WITH cte AS (
    SELECT
        a,
        b
    FROM foo
),

cte2 AS (
    SELECT
        a,
        c
    FROM bar
),

SELECT
    cte.a,
    cte.b,
    cte2.c
FROM cte
LEFT JOIN cte2
    ON cte.a = cte2.a;
```
1. `sqlfluff parse test.sql --dialect hive` and/or (`sqlfluff parse test.sql --dialect ansi`) 

### Dialect

Hive and ANSI (and possibly others could have the same issue).

### Version

```
-> % sqlfluff --version
sqlfluff, version 3.2.5
```
```
-> %python3 --version
Python 3.13.0
```

### Configuration

n/a

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by jcocozza** [comment]
I guess I'll add that I'd be willing to work on a PR to address the issue, but truth be told I haven't looked at this codebase and know very little about parsing etc. I probably would be of very little use. (Hence why I didn't check the box)

But I'd be happy to help if those considerations don't seem like huge blockers.
